### PR TITLE
[#530] Usage message customization

### PR DESCRIPTION
**Pull request by stechio**
Added support to flexible usage message building (see issue #530):
- [new] **customizable Help object** injectable through `CommandLine.helpFactory(IHelpFactory)`
- [new] **customizable usage message sections definition and ordering** through `List<ISectionRenderer> Help.sections()`
- [refactoring] **customizable usage message members** (headings, argument lists, ...) through their respective renderers (existing hard-coded rendering blocks have been encapsulated into the new objects and existing Help entry points have been wired back)

(no test added yet -- waiting for your considerations about this implementation)

### CODE REVIEW

**Comment by remkop** [comment]
Thanks! Limited internet connectivity now. Will try to review this week.

File: src/main/java/picocli/CommandLine.java [hunk 1]
```diff
@@ -3267,6 +3265,34 @@ static IDefaultValueProvider createDefaultValueProvider(IFactory factory, Class<
             catch (Exception ex) { throw new InitializationException("Could not instantiate " + cls + ": " + ex, ex); }
         }
     }
+
+    public interface IHelpFactory {
+        public Help createHelp(Object command);
+        public Help createHelp(CommandSpec commandSpec);
+        public Help createHelp(CommandSpec commandSpec, Help.Ansi ansi);
+        public Help createHelp(CommandSpec commandSpec, Help.ColorScheme colorScheme);
```
Comment (remkop) [inline src/main/java/picocli/CommandLine.java]: I prefer this interface to have just one method, that facilitates use with Java lambdas and Groovy closures.

File: src/main/java/picocli/CommandLine.java [hunk 2]
```diff
@@ -3267,6 +3265,34 @@ static IDefaultValueProvider createDefaultValueProvider(IFactory factory, Class<
...
+    }
+
+    public static abstract class BaseHelpFactory implements IHelpFactory {
```
Comment (remkop) [inline src/main/java/picocli/CommandLine.java]: I prefer to have as few classes and interfaces as possible. Just a DefaultHelpFactory is enough I think.

**Comment by remkop** [comment]
Perhaps a good way forward is to split this work. Can you provide a PR of more limited scope to address the first two points:

* allow injecting customizable Help object through `IHelpFactory`
* allow reordering and course-grained customization of help sections with `List<ISectionRenderer> Help.sections()`

We already agreed these would be good changes so I imagine this can be merged fairly quickly.

After that we can take another look at potential further improvements.

Thoughts?

**Comment by remkop** [comment]
Looking back at my initial feedback I realize now that it doesn't sound very friendly. Apologies, that was not intended. Are you still interested in helping to improve this area?

**Comment by remkop** [comment]
Looking back again my initial feedback doesn’t look that unfriendly.  Bit worried about not getting a response. I hope you’re okay. 
If you’re not interested in working on this further just give a shout.

**Comment by stechio** [comment]
Hi, sorry for my late reply: I had to move accordingly to my requirements, so I couldn't amend my proposal in the meantime.

I fully understand your observations (my redundant interfaces were an intermediate step toward the full modularization of the rendering elements), but at this point my implementation has nonetheless incompatibly diverged: I can keep alive my PR branch if you want to pick some of it, but I'm not gonna accomodate your request, sorry.

**Comment by remkop** [comment]
@stechio I wanted to thank you again for pointing out that in picocli-3.7/3.8 the usage help message customizability was not great. Thanks to your ideas, proposals and discussions, picocli-3.9.0 has just been [released](https://github.com/remkop/picocli/releases/tag/v3.9.0) with significant improvements in this area. Your name is [mentioned](https://github.com/remkop/picocli/releases/tag/v3.9.0#3.9.0-fixes) in the release notes to reflect this.

I believe [tickets like this one](https://github.com/remkop/picocli/issues/566) already prove how useful these improvements are.

I'm going to close this PR now since I think we have gotten as much value out of it as we can.

Again, many thanks, and please feel free to suggest further improvements!

**Pull request closed without merging**